File: tools-win/LOADER7VC/LOADER7VC/LOADER7VC.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Security.Cryptography;
using System.Text;

class Program
{
    static void Main(string[] args)
    {
        string EXEInput = null;

        // Check if there is a command line argument
        if (args.Length > 0)
        {
            // Use the provided command line argument as the file path
            EXEInput = args[0];

            // Validate provided file path
            if (!ValidateFilePath(EXEInput))
            {
                Console.WriteLine("Only LOADER.EXE or Dev7VM.EXE.");
                Environment.Exit(1);
            }
        }
        else
        {
            // Check if LOADER7.EXE or Dev7VM.EXE exists in the current directory
            string loader7Path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "LOADER7.EXE");
            string dev7VMPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Dev7VM.EXE");

            if (File.Exists(loader7Path))
            {
                EXEInput = loader7Path;
            }
            else if (File.Exists(dev7VMPath))
            {
                EXEInput = dev7VMPath;
            }
            else
            {
                Console.WriteLine("LOADER7.EXE or Dev7VM.EXE not found in the current directory.");
                Console.WriteLine("Press enter to exit");
                Console.ReadLine();
                Environment.Exit(1);
            }
        }

        try
        {
            // Display file information
            DisplayFileInfo(EXEInput, interactive: true);
        }
        catch (Exception ex)
        {
            Console.WriteLine("An error occurred: " + ex.Message);
            Environment.Exit(1);
        }
    }

    static void DisplayFileInfo(string filePath, bool interactive)
    {
        // Calculate and display MD5 checksum
        string md5Sum = CalculateMD5(filePath);

        // Get file size
        long fileSize = new FileInfo(filePath).Length;

        FileVersionInfo myFileVersionInfo = FileVersionInfo.GetVersionInfo(filePath);
        string EXEFilename = Path.GetFileName(filePath);

        Console.WriteLine("\nInfos about " + EXEFilename + ":\n");

        PrintIfNoInfos("Original Filename", myFileVersionInfo.OriginalFilename);
        PrintIfNoInfos("Description", myFileVersionInfo.FileDescription);
        PrintIfNoInfos("Version", myFileVersionInfo.FileVersion);
        PrintIfNoInfos("Productname", myFileVersionInfo.ProductName);
        PrintIfNoInfos("Internalname", myFileVersionInfo.InternalName);
        PrintIfNoInfos("Copyright", myFileVersionInfo.LegalCopyright);
        PrintIfNoInfos("Language", myFileVersionInfo.Language); // Language of the executable is displayed in the Language that your system is set to.

        Console.WriteLine("MD5 Checksum: " + md5Sum);
        Console.WriteLine($"File Size: {fileSize} Bytes");

        // Read the content of the EXE file
        string exeContent = File.ReadAllText(filePath);

        // Find the index of "NB10" in the content
        int nb10Index = exeContent.IndexOf("NB10");

        // Check if "NB10" is found in the content
        if (nb10Index != -1)
        {
            // Print everything after "NB10" to display the PDB File Path
            string afterNB10 = exeContent.Substring(nb10Index + 4);
            string pdbFilePath = GetPrintableString(afterNB10);
            PrintIfNoInfos("PDB File", pdbFilePath);
        }
        else
        {
            Console.WriteLine("PDB File reference not found in " + EXEFilename);
        }

        // Wait for user input if running in interactive mode
        if (interactive)
        {
            Console.WriteLine("\nPress any key to exit LOADER7VC");
            Console.ReadKey();
        }
    }

    static bool ValidateFilePath(string filePath)
    {
        // Add additional validation if needed
        return filePath.EndsWith("LOADER7.EXE", StringComparison.OrdinalIgnoreCase) ||
               filePath.EndsWith("Dev7VM.EXE", StringComparison.OrdinalIgnoreCase);
    }

    static void PrintIfNoInfos(string label, string value)
    {
        if (!string.IsNullOrEmpty(value))
        {
            Console.WriteLine($"{label}: {value}");
        }
        else
        {
            Console.WriteLine($"{label}: Not available in LOADER7.EXE/Dev7VM.EXE.");
        }
    }

    static string CalculateMD5(string filePath)
    {
        using (var md5 = MD5.Create())
        {
            using (var stream = File.OpenRead(filePath))
            {
                byte[] hash = md5.ComputeHash(stream);
                return BitConverter.ToString(hash).Replace("-", "").ToLower();
            }
        }
    }

    static string GetPrintableString(string input)
    {
        // Remove non-printable characters from the input string
        StringBuilder printableString = new StringBuilder();

        foreach (char c in input)
        {
            if (Char.IsControl(c) && !Char.IsWhiteSpace(c))
            {
                // Skip non-printable characters
            }
            else
            {
                printableString.Append(c);
            }
        }

        return printableString.ToString();
    }
}

File: tools-win/LOADER7VC/LOADER7VC/LOADER7VC.cpp
#include <windows.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <Shlwapi.h>
#include <iomanip>

#pragma comment(lib, "Version.lib")
#pragma comment(lib, "Shlwapi.lib")

void PrintHeader() {
    std::string versionNumber = "v1.5-testing";
    std::cout << "============================\n";
    std::cout << "LOADER7VC " << versionNumber << "\n";
    std::cout << "============================\n";
}

bool ValidateFilePath(const std::string& filePath) {
    return StrStrIA(filePath.c_str(), "Loader7.exe") || StrStrIA(filePath.c_str(), "Dev7VM.EXE");
}

void PrintVersionInfo(const std::string& filename) {
    DWORD dummy;
    DWORD size = GetFileVersionInfoSizeA(filename.c_str(), &dummy);
    if (size == 0) {
        std::cout << "No version info found.\n";
        return;
    }

    std::vector<char> data(size);
    if (!GetFileVersionInfoA(filename.c_str(), 0, size, data.data())) {
        std::cout << "GetFileVersionInfo failed.\n";
        return;
    }

    VS_FIXEDFILEINFO* verInfo = nullptr;
    UINT len = 0;
    if (VerQueryValueA(data.data(), "\\", reinterpret_cast<LPVOID*>(&verInfo), &len)) {
        std::cout << "Version: "
            << HIWORD(verInfo->dwFileVersionMS) << "."
            << LOWORD(verInfo->dwFileVersionMS) << "."
            << HIWORD(verInfo->dwFileVersionLS) << "."
            << LOWORD(verInfo->dwFileVersionLS) << "\n";
    }

    struct StringInfo {
        LPCSTR name;
        const char* label;
    };

    StringInfo fields[] = {
        {"OriginalFilename", "Original Filename"},
        {"FileDescription",  "Description"},
        {"ProductName",      "Productname"},
        {"InternalName",     "Internalname"},
        {"LegalCopyright",   "Copyright"},
        {"Language",         "Language"},
    };

    LPVOID pBuffer = nullptr;
    UINT cbTranslate = 0;
    if (VerQueryValueA(data.data(), "\\VarFileInfo\\Translation", &pBuffer, &cbTranslate)) {
        WORD* langCode = (WORD*)pBuffer;
        char subBlock[100];
        for (auto& field : fields) {
            sprintf_s(subBlock, "\\StringFileInfo\\%04x%04x\\%s", langCode[0], langCode[1], field.name);
            LPVOID val = nullptr;
            UINT size;
            if (VerQueryValueA(data.data(), subBlock, &val, &size)) {
                std::cout << field.label << ": " << (char*)val << "\n";
            }
            else {
                std::cout << field.label << ": Not available.\n";
            }
        }
    }
}

std::string GetPrintableString(const std::string& input) {
    std::string output;
    for (char c : input) {
        if (isprint(static_cast<unsigned char>(c)) || isspace(static_cast<unsigned char>(c))) {
            output += c;
        }
    }
    return output;
}

void DisplayFileInfo(const std::string& filepath) {
    std::ifstream file(filepath, std::ios::binary);
    if (!file) {
        std::cerr << "Failed to open file.\n";
        return;
    }

    // Read entire file into memory (only for small files)
    std::string content((std::istreambuf_iterator<char>(file)),
        std::istreambuf_iterator<char>());

    std::cout << "\nInfos about " << filepath << ":\n\n";

    PrintVersionInfo(filepath);

    std::ifstream sizeFile(filepath, std::ios::binary | std::ios::ate);
    std::cout << "File Size: " << sizeFile.tellg() << " Bytes\n";

    size_t nb10Index = content.find("NB10");
    if (nb10Index != std::string::npos) {
        std::string afterNB10 = content.substr(nb10Index + 4);
        std::string pdbPath = GetPrintableString(afterNB10);
        std::cout << "PDB File: " << pdbPath << "\n";
    }
    else {
        std::cout << "PDB File reference not found in " << filepath << "\n";
    }

    std::cout << "\nPress Enter to exit LOADER7VC...";
    std::cin.get();
}

int main(int argc, char* argv[]) {
    PrintHeader();

    std::string exePath;

    if (argc > 1) {
        exePath = argv[1];
        if (!ValidateFilePath(exePath)) {
            std::cerr << "Only Loader7.exe or Dev7VM.EXE are allowed.\n";
            return 1;
        }
    }
    else {
        char buffer[MAX_PATH];
        GetModuleFileNameA(NULL, buffer, MAX_PATH);
        std::string dirPath = buffer;
        dirPath = dirPath.substr(0, dirPath.find_last_of("\\/"));

        std::string loader7 = dirPath + "\\Loader7.exe";
        std::string dev7vm = dirPath + "\\Dev7VM.EXE";

        if (PathFileExistsA(loader7.c_str())) {
            exePath = loader7;
        }
        else if (PathFileExistsA(dev7vm.c_str())) {
            exePath = dev7vm;
        }
        else {
            std::cerr << "Loader7.exe or Dev7VM.EXE not found in the current directory.\n";
            std::cout << "Press Enter to exit.\n";
            std::cin.get();
            return 1;
        }
    }

    DisplayFileInfo(exePath);
    return 0;
}
